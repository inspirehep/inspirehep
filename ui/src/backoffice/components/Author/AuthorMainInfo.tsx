import React from 'react';
import { Map } from 'immutable';
import classNames from 'classnames';
import ContentBox from '../../../common/components/ContentBox';
import { Ids } from '../Links';

type AuthorMainInfoProps = {
  data: any; // TODO: define proper type for data
};

const hasOrcidId = (data: any): boolean => {
  return data
    .get('ids')
    ?.find(
      (id: { get: (arg0: string) => string }) => id.get('schema') === 'ORCID'
    );
};

const filterOrcidIds = (data: any) => {
  return data
    .get('ids')
    ?.filter((id: Map<string, any>) => id.get('schema') === 'ORCID');
};

const AuthorMainInfo = ({ data }: AuthorMainInfoProps) => {
  const name = data.getIn(['name', 'value']);
  const prefferedName = data.getIn(['name', 'preferred_name']);
  const nativeNames = data.getIn(['name', 'native_names']);
  const nameVariants = data.getIn(['name', 'name_variants']);
  const status = data.get('status');
  const orcidIdExists = hasOrcidId(data);
  const orcidIds = filterOrcidIds(data);

  return (
    <ContentBox fullHeight={false} className="md-pb3 mb3">
      <h2>{name}</h2>
      {prefferedName && (
        <p>
          <b>Preferred name:</b> {prefferedName}
        </p>
      )}
      {nativeNames && (
        <p>
          <b>Native names:</b> {nativeNames.join('; ')}
        </p>
      )}
      {nameVariants && (
        <p>
          <b>Name variants:</b> {nameVariants.join('; ')}
        </p>
      )}
      {status && (
        <p className={classNames({ mb0: !orcidIdExists })}>
          <b>Status:</b> {status}
        </p>
      )}
      {orcidIdExists && <Ids ids={orcidIds} noIcon />}
    </ContentBox>
  );
};

export default AuthorMainInfo;
