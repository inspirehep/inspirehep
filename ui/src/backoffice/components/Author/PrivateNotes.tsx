import React from 'react';
import ContentBox from '../../../common/components/ContentBox';

type PrivateNotesProps = {
  privateNotes: { get: (arg0: string) => {} | null | undefined }[]; // TODO: define proper type for privateNotes
};

const PrivateNotes = ({ privateNotes }: PrivateNotesProps) => (
  <ContentBox className="mb3" fullHeight={false} subTitle="Notes">
    <i>
      {privateNotes.map((note) => {
        const value = note?.get('value');
        return (
          <p className="mb0" key={value as string}>
            &quot;{value}&quot;
          </p>
        );
      })}
    </i>
  </ContentBox>
);

export default PrivateNotes;
