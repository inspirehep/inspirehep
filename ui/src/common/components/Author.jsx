import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import { Map } from 'immutable';
import { Tooltip } from 'antd';

import AuthorAffiliationList from './AuthorAffiliationList';
import { AUTHORS } from '../routes';

class Author extends Component {
  getAuthorName() {
    const { author } = this.props;
    if (author.has('first_name')) {
      const firstName = author.get('first_name');
      const lastName = author.get('last_name', '');
      return `${firstName} ${lastName}`;
    }
    return author.get('full_name');
  }

  getAuthorRecid() {
    const { author } = this.props;
    const ref = author.getIn(['record', '$ref']);

    if (!ref) {
      return null;
    }

    const urlParts = ref.split('/');
    return urlParts[urlParts.length - 1];
  }

  renderRoleSuffix() {
    const { author } = this.props;
    const roles = author.get('inspire_roles', []);

    if (roles.indexOf('editor') > -1) {
      return <Tooltip title="editor">(ed.)</Tooltip>;
    }

    return null;
  }

  renderAffiliationsList() {
    const { author } = this.props;
    const affiliations = author.get('affiliations');
    return (
      affiliations && (
        <span className="pl1 secondary-color">
          (
          <AuthorAffiliationList
            affiliations={affiliations.map(affiliation =>
              affiliation.get('value')
            )}
          />
          )
        </span>
      )
    );
  }

  renderAuthorName() {
    const recid = this.getAuthorRecid();
    if (recid != null) {
      return <Link to={`${AUTHORS}/${recid}`}>{this.getAuthorName()}</Link>;
    }
    return <span>{this.getAuthorName()}</span>;
  }

  render() {
    return (
      <div className="di">
        {this.renderAuthorName()}
        {this.renderAffiliationsList()}
        {this.renderRoleSuffix()}
      </div>
    );
  }
}

Author.propTypes = {
  author: PropTypes.instanceOf(Map).isRequired,
};

export default Author;
