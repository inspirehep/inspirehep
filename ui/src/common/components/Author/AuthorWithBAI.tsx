import React from 'react';
import PropTypes from 'prop-types';
import { Map } from 'immutable';
import { Link } from 'react-router-dom';
import { Tooltip } from 'antd';

import { getAuthorName, getLiteratureSearchUrlForAuthorBAI } from '../../utils';
import ExternalLink from '../ExternalLink.tsx';
import { SUBMISSIONS_AUTHOR } from '../../routes';

function renderCreateProfileTooltipMessage(author) {
  return (
    <>
      <div>The author does not have an INSPIRE profile</div>
      <div>
        <ExternalLink href={`${SUBMISSIONS_AUTHOR}?bai=${author.get('bai')}`}>
          Create profile
        </ExternalLink>
      </div>
    </>
  );
}

function AuthorWithBAI({ author }) {
  return (
    <Tooltip title={renderCreateProfileTooltipMessage(author)}>
      <Link
        style={{ textDecoration: 'underline dashed' }}
        to={getLiteratureSearchUrlForAuthorBAI(author.get('bai'))}
      >
        {getAuthorName(author)}
      </Link>
    </Tooltip>
  );
}

AuthorWithBAI.propTypes = {
  author: PropTypes.instanceOf(Map).isRequired,
};

export default AuthorWithBAI;
