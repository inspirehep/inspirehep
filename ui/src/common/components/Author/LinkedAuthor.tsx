import React from 'react';
import PropTypes from 'prop-types';
import { Map } from 'immutable';
import { Link } from 'react-router-dom';
import { getAuthorName, getRecordIdFromRef } from '../../utils';
import { AUTHORS } from '../../routes';

function LinkedAuthor({ author }) {
  return (
    <Link
      data-test-id="author-link"
      to={`${AUTHORS}/${getRecordIdFromRef(author.getIn(['record', '$ref']))}`}
    >
      {getAuthorName(author)}
    </Link>
  );
}

LinkedAuthor.propTypes = {
  author: PropTypes.instanceOf(Map).isRequired,
};

export default LinkedAuthor;
