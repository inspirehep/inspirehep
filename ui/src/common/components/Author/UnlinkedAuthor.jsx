import React from 'react';
import PropTypes from 'prop-types';
import { Map } from 'immutable';
import { getAuthorName } from '../../utils';

function UnlinkedAuthor({ author }) {
  return <span>{getAuthorName(author)}</span>;
}

UnlinkedAuthor.propTypes = {
  author: PropTypes.instanceOf(Map).isRequired,
};

export default UnlinkedAuthor;
