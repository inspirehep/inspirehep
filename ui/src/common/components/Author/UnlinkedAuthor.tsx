import React from 'react';
import { Map } from 'immutable';
import { getAuthorName } from '../../utils';

type Props = {
    author: $TSFixMe; // TODO: PropTypes.instanceOf(Map)
};

function UnlinkedAuthor({ author }: Props) {
  return <span>{getAuthorName(author)}</span>;
}

export default UnlinkedAuthor;
