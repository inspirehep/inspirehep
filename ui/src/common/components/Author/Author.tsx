import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Map } from 'immutable';
import { Tooltip } from 'antd';

import AffiliationList from '../AffiliationList';
import UnlinkedAuthor from './UnlinkedAuthor';
import AuthorWithBAI from './AuthorWithBAI';
import LinkedAuthor from './LinkedAuthor';

class Author extends Component {
  renderRoleSuffix() {
    const { author } = this.props;
    const roles = author.get('inspire_roles', []);

    if (roles.indexOf('editor') > -1) {
      return <Tooltip title="editor">(ed.)</Tooltip>;
    }

    return null;
  }

  renderAffiliationsList() {
    const { author } = this.props;
    const affiliations = author.get('affiliations');
    return (
      affiliations && (
        <span className="pl1 secondary-color">
          (
          <AffiliationList affiliations={affiliations} />
          )
        </span>
      )
    );
  }

  renderAuthorName() {
    const { author } = this.props;
    if (author.has('record')) {
      return <LinkedAuthor author={author} />;
    }
    if (author.has('bai')) {
      return <AuthorWithBAI author={author} />;
    }
    return <UnlinkedAuthor author={author} />;
  }

  render() {
    return (
      <div className="di">
        {this.renderAuthorName()}
        {this.renderAffiliationsList()}
        {this.renderRoleSuffix()}
      </div>
    );
  }
}

Author.propTypes = {
  author: PropTypes.instanceOf(Map).isRequired,
};

export default Author;
