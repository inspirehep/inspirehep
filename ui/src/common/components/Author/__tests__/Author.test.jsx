import React from 'react';
import { fromJS } from 'immutable';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

import Author from '..';

describe('Author', () => {
  it('renders unlinkedAuthor with affiliations', () => {
    const author = fromJS({
      full_name: 'Name, Full, Jr.',
      first_name: 'Full, Jr.',
      last_name: 'Name',
      affiliations: [
        {
          value: 'Affiliation',
        },
      ],
    });
    const { getByText } = render(<Author author={author} recordId={12345} />);
    expect(getByText('Full Name, Jr.')).toBeInTheDocument();
    expect(getByText('Affiliation')).toBeInTheDocument();
  });

  it('renders editor', () => {
    const author = fromJS({
      full_name: 'Name, Full',
      first_name: 'Full',
      last_name: 'Name',
      inspire_roles: ['editor'],
      affiliations: [
        {
          value: 'Affiliation',
        },
      ],
    });
    const { getByText } = render(<Author recordId={12345} author={author} />);
    expect(getByText('Full Name')).toBeInTheDocument();
    expect(getByText('Affiliation')).toBeInTheDocument();
  });

  it('renders supervisor', () => {
    const author = fromJS({
      full_name: 'Name, Full',
      inspire_roles: ['supervisor'],
    });
    const { getByText } = render(<Author recordId={12345} author={author} />);
    expect(getByText('Name, Full')).toBeInTheDocument();
  });

  it('renders linked author', () => {
    const author = fromJS({
      full_name: 'Name, Full',
      record: {
        $ref: 'https://beta.inspirehep.net/api/authors/12345',
      },
      bai: 'Full.Name.1',
    });
    const { getByRole } = render(
      <MemoryRouter>
        <Author recordId={12345} author={author} />
      </MemoryRouter>
    );
    const link = getByRole('link', {
      name: 'Name, Full',
    });
    expect(link).toHaveAttribute('href', '/authors/12345');
  });

  it('renders unlinked author with bai', () => {
    const author = fromJS({
      full_name: 'Name, Full',
      bai: 'Full.Name.1',
    });
    const { getByRole } = render(
      <MemoryRouter>
        <Author recordId={12345} author={author} />
      </MemoryRouter>
    );
    const link = getByRole('link', {
      name: 'Name, Full',
    });
    expect(link).toHaveAttribute('href', '/literature?q=a%20Full.Name.1');
  });
});
