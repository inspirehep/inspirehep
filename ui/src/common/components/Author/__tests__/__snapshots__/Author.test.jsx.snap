// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Author renders editor 1`] = `
<div
  className="di"
>
  <UnlinkedAuthor
    author={
      Immutable.Map {
        "full_name": "Name, Full",
        "first_name": "Full",
        "last_name": "Name",
        "inspire_roles": Immutable.List [
          "editor",
        ],
        "affiliations": Immutable.List [
          Immutable.Map {
            "value": "Affiliation",
          },
        ],
      }
    }
  />
  <span
    className="pl1 secondary-color"
  >
    (
    <AffiliationList
      affiliations={
        Immutable.List [
          Immutable.Map {
            "value": "Affiliation",
          },
        ]
      }
      separator=" and "
    />
    )
  </span>
  <Tooltip
    arrowPointAtCenter={false}
    autoAdjustOverflow={true}
    mouseEnterDelay={0.1}
    mouseLeaveDelay={0.1}
    placement="top"
    title="editor"
  >
    (ed.)
  </Tooltip>
</div>
`;

exports[`Author renders linked author 1`] = `
<div
  className="di"
>
  <LinkedAuthor
    author={
      Immutable.Map {
        "full_name": "Name, Full",
        "record": Immutable.Map {
          "$ref": "https://beta.inspirehep.net/api/authors/12345",
        },
        "bai": "Full.Name.1",
      }
    }
  />
</div>
`;

exports[`Author renders supervisor 1`] = `
<div
  className="di"
>
  <UnlinkedAuthor
    author={
      Immutable.Map {
        "full_name": "Name, Full",
        "inspire_roles": Immutable.List [
          "supervisor",
        ],
      }
    }
  />
</div>
`;

exports[`Author renders unlinked author with bai 1`] = `
<div
  className="di"
>
  <AuthorWithBAI
    author={
      Immutable.Map {
        "full_name": "Name, Full",
        "bai": "Full.Name.1",
      }
    }
  />
</div>
`;

exports[`Author renders unlinkedAuthor with affiliations 1`] = `
<div
  className="di"
>
  <UnlinkedAuthor
    author={
      Immutable.Map {
        "full_name": "Name, Full, Jr.",
        "first_name": "Full, Jr.",
        "last_name": "Name",
        "affiliations": Immutable.List [
          Immutable.Map {
            "value": "Affiliation",
          },
        ],
      }
    }
  />
  <span
    className="pl1 secondary-color"
  >
    (
    <AffiliationList
      affiliations={
        Immutable.List [
          Immutable.Map {
            "value": "Affiliation",
          },
        ]
      }
      separator=" and "
    />
    )
  </span>
</div>
`;
